x-pack/legacy/plugins/apm/server/lib/helpers/get_apm_indices.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Server } from 'hapi';
+import { isObject } from 'lodash';
+import { getSavedObjectsClient } from './saved_objects_client';
+
+interface UiIndices {
+  apm_oss: {
+    sourcemapIndices?: string;
+    errorIndices?: string;
+    onboardingIndices?: string;
+    spanIndices?: string;
+    transactionIndices?: string;
+    metricsIndices?: string;
+    apmAgentConfigurationIndex?: string;
+  };
+}
+
+export async function getApmIndices(server: Server) {
+  const config = server.config();
+
+  const savedObjectsClient = getSavedObjectsClient(server, 'data');
+  // eslint-disable-next-line @typescript-eslint/camelcase
+  let apm_oss: UiIndices['apm_oss'] = {};
+  try {
+    const apmUiIndices = await savedObjectsClient.get<UiIndices>(
+      'apm-ui-indices',
+      'apm-ui-indices'
+    );
+    // eslint-disable-next-line @typescript-eslint/camelcase
+    apm_oss = apmUiIndices.attributes.apm_oss;
+  } catch (err) {
+    server.log('error', err);
+  }
+  return {
+    apm_oss: {
+      sourcemapIndices:
+        apm_oss.sourcemapIndices ||
+        config.get<string>('apm_oss.sourcemapIndices'),
+      errorIndices:
+        apm_oss.errorIndices || config.get<string>('apm_oss.errorIndices'),
+      onboardingIndices:
+        apm_oss.onboardingIndices ||
+        config.get<string>('apm_oss.onboardingIndices'),
+      spanIndices:
+        apm_oss.spanIndices || config.get<string>('apm_oss.spanIndices'),
+      transactionIndices:
+        apm_oss.transactionIndices ||
+        config.get<string>('apm_oss.transactionIndices'),
+      metricsIndices:
+        apm_oss.metricsIndices || config.get<string>('apm_oss.metricsIndices'),
+      apmAgentConfigurationIndex:
+        apm_oss.apmAgentConfigurationIndex ||
+        config.get<string>('apm_oss.apmAgentConfigurationIndex')
+    }
+  };
+}
+
+function flatValues<T = any>(obj: Record<string, any>, deepValues = []): T[] {
```

Comment:
I'm having a bit of a hard time understanding what this function does. I guess it's getting all the values from `apm_oss`? Would it be better to just explicitly define that here? Seems like `deepValues` is not being used either? e.g.:

```typescript
return Object.values(apmIndices.apm_oss);
```

Reply:
the purpose of `flatValues` as opposed to `Object.values(apmIndices.apm_oss);` is to keep `getApmIndicesList` generic. If we add a new index configuration, e.g. `xpack.apm.ui.someNewIndex`, then it gets included in `getApmIndicesList` without having to change that function. I open to changing this, but that was my original motivation.